main/lsp/ExtractVariable.cc

```diff
@@ -0,0 +1,239 @@
+#include "main/lsp/ExtractVariable.h"
+#include "ast/treemap/treemap.h"
+
+using namespace std;
+
+namespace sorbet::realmain::lsp {
+
+class ExtractVariableWalk {
```

Comment:
This tree walk has the core logic of the code action. It has 2 aims:
- check if the selection is a valid selection to extract
- find which "scope" the selection is inside

The first is mostly straightforward. In `preTransformExpression`, if we find an node with the exact same `loc`, and it's not a node we can't extract (like `return`), we can allow extracting that selection. The one case that requires some extra handling is when we have identifiers (which are usually fine to extract) in positions that disallow extraction. For example, in `def foo(x); end; y = 1`, extracting both `x` and `y` would be invalid. To handle this, during the tree walk, we call `skipLoc` on locs that represent those invalid locations. For example, in `preTransformMethodDef`, we loop through each arg and call `skipLoc` for each of the arguments' loc. This means that if the selection is one of the arguments, `foundExactMatch` will still return false since it's loc will be in `skippedLocs`.

The second is bit more complicated to do. Naively, we could just look for the deepest `InsSeq` that contains the selection and say that's the scope where it's defined. The problem with that approach is that we omit for `InsSeq`s for single line scope. For example, `def foo(x); x; end` will have the body of the method as a reference to a `UnresolvedIdent` directly rather than an `InsSeq` that contains the identifier. To handle this, we also need to walk over all the nodes that could introduce a "scope" (`if`, `methodDef`, `classDef` etc.). We still match on `InsSeq`s in addition, so that if scope does have a `InsSeq`, we'll use that instead.